src/main/scala/firrtl/passes/Passes.scala

```diff
@@ -247,23 +249,204 @@ object Legalize extends Pass {
+    */
+  private def onExpression(e: Expression)
+                          (implicit renames: RenameMap, ns: Namespace, scope: Option[ModuleName],
+                           iToM: mutable.Map[ComponentName, ModuleName]): Expression = e match {
+    case wsf@ WSubField(wr@ WRef(name, _, InstanceKind, _), port, _, _) =>
+      val subInst = ComponentName(name, scope.get)
+      val subModule = iToM(subInst)
+      val subPort = ComponentName(port, subModule)
+
+      val wrx = wr.copy(
+        name = renames.get(subInst).orElse(Some(Seq(subInst))).get.head.name,
+        tpe = onType(wr.tpe)(renames, ns, Some(subModule)))
 
-  def verilogRenameE(e: Expression): Expression = e match {
-    case ex: WRef => ex copy (name = verilogRenameN(ex.name))
-    case ex => ex map verilogRenameE
+      wsf.copy(
+        expr = wrx,
+        name = renames.get(subPort).orElse(Some(Seq(subPort))).get.head.name)
+    case wr: WRef => wr.copy(name=onName(wr.name))
+    case ex       => ex.map(onExpression)
   }
 
-  def verilogRenameS(s: Statement): Statement =
-    s map verilogRenameS map verilogRenameE map verilogRenameN
+  /** Rename a [[Statement]] to respect existing renames and avoid keyword collisions
+    * $implicitRename
+    * $implicitNamespace
+    * $implicitScope
+    * @return a [[Statement]] without keyword conflicts
+    */
+  private def onStatement(s: Statement)
+                         (implicit renames: RenameMap, ns: Namespace, scope: Option[ModuleName],
+                          iToM: mutable.Map[ComponentName, ModuleName]): Statement = s match {
+    case wdi: WDefInstance =>
+      val subModule = ModuleName(wdi.module, scope.get.circuit)
+      val modulex = renames.get(subModule).orElse(Some(Seq(subModule))).get.head.name
+      val wdix = wdi.copy(module = modulex,
+                          name = onName(wdi.name),
+                          tpe = onType(wdi.tpe)(renames, ns, Some(ModuleName(modulex, scope.get.circuit))))
+      iToM(ComponentName(wdi.name, scope.get)) = ModuleName(wdix.module, scope.get.circuit)
+      wdix
+    case s => s
+        .map(onStatement)
+        .map(onExpression)
+        .map(onName)
+  }
+
+  /** Rename a [[Port]] to avoid keyword collisions
+    * $implicitRename
+    * $implicitNamespace
+    * $implicitScope
+    * @return a [[Port]] without keyword conflicts
+    */
+  private def onPort(p: Port)(implicit renames: RenameMap, ns: Namespace, scope: Option[ModuleName]): Port =
+    p.copy(name = onName(p.name))
+
+  /** Rename a [[DefModule]] and it's internals (ports and statements) to fix keyword collisions and update instance
+    * references to respect previous renames
+    * @param renames a [[RenameMap]]
+    * @param circuit the enclosing [[CircuitName]]
+    * @return a [[DefModule]] without keyword conflicts
+    */
+  private def onModule(renames: RenameMap, circuit: CircuitName)(m: DefModule): DefModule = {
+    implicit val moduleNamespace = Namespace(m)
```

Comment:
IntelliJ feels that these lines here should have explicit types

Reply:
I'm less enthusiastic about the wills of our IntelliJ overlords... I wonder why IntelliJ wants this?

Did some digging... found one thread about this. We could turn this off in IntelliJ, but this may be mandatory in Dotty. I'll change these. Type annotations are usually always good. https://blog.jetbrains.com/scala/2016/10/05/beyond-code-style/#comment-198824

Reply:
Thanks for the link, That's a decent discussion of the issues.